lib/plugins/aws/invokeLocal/index.js

```diff
@@ -127,10 +127,11 @@ class AwsInvokeLocal {
     });
   }
 
-  getCredentialEnvVars() {
+  async getCredentialEnvVars() {
     const credentialEnvVars = {};
     const { credentials } = this.provider.getCredentials();
     if (credentials) {
+      await credentials.getPromise().catch(() => null); // best effort
```

Comment:
An AWS credentials provider isn't setup until `get` (or `getPromise` if you don't want to deal with callbacks) is called. Until then, accessing `accessKeyId`, etc. will get you either `undefined` or stale credentials.

If you don't have your system configured to resolve valid credentials, however, (like in a good few of the tests) this will throw because the credentials could not be resolved.

Normally throwing is the right thing - it is for the awsProvider, because it means you can't authenticate with AWS.

However, it looks like (for whatever reason, L150: `return _.merge(providerEnvVars, functionEnvVars);`) `invokeLocal` is doing its _own_ credential resolution on top of the awsProvider. In this case we gracefully fail, so that not having valid provider credentials doesn't prevent you from invoking with valid `functionEnvVars`.

Reply:
In light of that I believe this should be handled internally by our `credentials` class. Expecting it's consumers to do such calls to make credentials effective doesn't look as right pattern

Reply:
Hmm- but this is how AWS credentials are _supposed_ to work, and are used within the aws-sdk itself.

They're a class that you call (`get`, or `refresh`) on (etc.) to ensure you have the latest credentials prior to making a request. [(docs)](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Credentials.html).

It's also currently how they work within the framework at the moment. If you use any of the asynchronous approaches (like roles), then your access key / secret token is `undefined` or potentially stale until you call `refresh` or `get`, etc. - we're returning a `credentials` object. You can see workarounds for this in the original tests, with all the resolved credentials having to test for the private field `role`.

While we _could_ making a breaking change to `getCredentials` that makes it asynchronous and returns resolved credentials, it's main usage seems to be to return a credentials object that can be consumed by an AWS service, which will automatically call `refresh`.